js/client-chat.js

```diff
@@ -208,29 +208,39 @@
 		// highlight
 
 		getHighlight: function (message) {
-			var highlights = Dex.prefs('highlights') || [];
-			if (!app.highlightRegExp) {
-				try {
-					this.updateHighlightRegExp(highlights);
-				} catch (e) {
-					// If the expression above is not a regexp, we'll get here.
-					// Don't throw an exception because that would prevent the chat
-					// message from showing up, or, when the lobby is initialising,
-					// it will prevent the initialisation from completing.
-					return false;
-				}
+			if (Array.isArray(Dex.prefs('highlights'))) {
+				this.convertHighlights(Dex.prefs('highlights'));
 			}
+			var id = Config.server.id + '#' + this.id;
+			var allHighlights = Dex.prefs('highlights') || {};
+			var roomHighlights = allHighlights[id] || [];
+			var globalHighlights = allHighlights['global'] || [];
+			var highlights = roomHighlights.concat(globalHighlights);
 			if (!Dex.prefs('noselfhighlight') && app.user.nameRegExp) {
 				if (app.user.nameRegExp.test(message)) return true;
 			}
-			return ((highlights.length > 0) && app.highlightRegExp.test(message));
+			var highlighRegExp;
+			try {
+				highlighRegExp = this.getHighlightRegExp(highlights);
+			} catch (e) {
+				// If the expression above is not a regexp, we'll get here.
+				// Don't throw an exception because that would prevent the chat
+				// message from showing up, or, when the lobby is initialising,
+				// it will prevent the initialisation from completing.
+				return false;
+			}
+			return ((highlights.length > 0) && highlighRegExp.test(message));
+		},
+		convertHighlights: function (highlights) {
+			var newHLs = {global: highlights};
+			Dex.prefs('highlights', newHLs);
 		},
-		updateHighlightRegExp: function (highlights) {
+		getHighlightRegExp: function (highlights) {
 			// Enforce boundary for match sides, if a letter on match side is
 			// a word character. For example, regular expression "a" matches
 			// "a", but not "abc", while regular expression "!" matches
 			// "!" and "!abc".
-			app.highlightRegExp = new RegExp('(?:\\b|(?!\\w))(?:' + highlights.join('|') + ')(?:\\b|(?!\\w))', 'i');
+			return new RegExp('(?:\\b|(?!\\w))(?:' + highlights.join('|') + ')(?:\\b|(?!\\w))', 'i');
```

Comment:
You've removed the caching...

Reply:
Oh, I didn't realize that's what it was for, lol. Fixed here: https://github.com/Zarel/Pokemon-Showdown-Client/pull/1334/commits/a4d110a98c758d17cc408e3936ed2422ff080b1f.